TLSv1.3 earlydata support for curl (#15211)
curl/curl: docs/libcurl/opts/CURLINFO_EARLYDATA_SENT_T.md

---
c: Copyright (C) Daniel Stenberg, <daniel@haxx.se>, et al.
SPDX-License-Identifier: curl
Title: CURLINFO_EARLYDATA_SENT_T
Section: 3
Source: libcurl
See-also:
  - curl_easy_getinfo (3)
  - curl_easy_setopt (3)

> Protocol:
>   - All

Reviewer: Since early data is only for TLS protocols and the feature only works for GnuTLS for now, I propose:

```suggestion
Protocol:
  - TLS
TLS-backend:
  - GnuTLS
```

Author: ack

Added-in: 8.11.0
---

# NAME

> CURLINFO_EARLYDATA_SENT_T - get the number of bytes sent as TLS earlydata

Reviewer: RFC 8446 speaks of it as **early data**, that is using two words. So I figure we should as well:

```suggestion
CURLINFO_EARLYDATA_SENT_T - get the number of bytes sent as TLS early data
```

Author: ack

# SYNOPSIS

~~~c
#include <curl/curl.h>

CURLcode curl_easy_getinfo(CURL *handle, CURLINFO_EARLYDATA_SENT_T,
                           curl_off_t *amount);
~~~

# DESCRIPTION

> Pass a pointer to a *curl_off_t* to receive the total amount of bytes that

Reviewer: ```suggestion
Pass a pointer to an *curl_off_t* to receive the total amount of bytes that
```

Author: not sure why, but I'll change it.

were sent to the server as TLSv1.3 earlydata. When no TLS early
data is used, this reports 0.

TLS earlydata is only attempted when CURLSSLOPT_EARLYDATA is set for the
transfer. In addition, it is only used by libcurl when a TLS session exists
that announces support.

> Note that the amount is **negative** when the sent data was rejected

Reviewer: ```suggestion
The amount is **negative** when the sent data was rejected
```

Author: ack

> by the server. TLS allows a server that announces support for earlydata to

Reviewer: ```suggestion
by the server. TLS allows a server that announces support for early data to
```

Author: ack, and some other places.
